src/awareness/awareness-state.ts

```diff
@@ -0,0 +1,295 @@
+	 */
+	public onStateChange( callback: ( newState: EnhancedState< State >[] ) => void ): () => void {
+		this.stateSubscriptions.push( callback );
+
+		return () => {
+			this.stateSubscriptions = this.stateSubscriptions.filter( cb => cb !== callback );
+		};
+	}
+
+	/**
+	 * Set a local state field on an awareness document with debounce. See caveats
+	 * of this.setLocalStateField.
+	 */
+	public setLocalStateFieldWithDebounce< FieldName extends string & keyof State >(
+		field: FieldName,
+		value: State[ FieldName ],
+		debounceMs: number
+	): void {
+		if ( this.debounceTimeouts.has( field ) ) {
+			clearTimeout( this.debounceTimeouts.get( field ) );
+		}
+
+		// Store the debounced value locally to merge with the actual state. We do
+		// this so that when we read our own state, we get the latest value even
+		// when it hasn't yet been set on the awareness instance.
+		// eslint-disable-next-line security/detect-object-injection
+		this.myDebouncedState[ field ] = value;
+
+		this.debounceTimeouts.set(
+			field,
+			setTimeout( () => {
+				// eslint-disable-next-line @typescript-eslint/no-dynamic-delete, security/detect-object-injection
+				delete this.myDebouncedState[ field ];
+				this.setLocalStateField( field, value );
+			}, debounceMs )
+		);
+	}
+
+	/**
+	 * Set the current user's connection status as awareness state.
+	 */
+	public setConnectionStatus( isConnected: boolean ): void {
+		if ( isConnected ) {
+			this.disconnectedUsers.delete( this.clientID );
+		} else {
+			this.disconnectedUsers.add( this.clientID );
+		}
+
+		this.updateSubscribers( true /* force update */ );
+	}
+
+	/**
+	 * Update all subscribed listeners with the latest awareness state.
+	 */
+	protected updateSubscribers( forceUpdate = false ): void {
+		if ( ! this.stateSubscriptions.length ) {
+			return;
+		}
+
+		const states = this.getStates();
+
+		this.seenStates = new Map< number, State >( [
+			...this.seenStates.entries(),
+			...states.entries(),
+		] );
+
+		const updatedStates = new Map< number, EnhancedState< State > >(
+			[ ...this.disconnectedUsers, ...states.keys() ]
+				.filter( clientId => this.seenStates.has( clientId ) )
+				.map( clientId => {
+					// The filter above ensures that seenStates has the clientId.
+					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+					const rawState: State = this.seenStates.get( clientId )!;
+
+					const isConnected = ! this.disconnectedUsers.has( clientId );
+					const isMe = clientId === this.clientID;
+					const myState: Partial< State > = isMe ? this.myDebouncedState : {};
+					const state: EnhancedState< State > = {
+						...rawState,
+						...myState,
+						clientId,
+						isConnected,
+						isMe,
+					};
+
+					return [ clientId, state ];
+				} )
+		);
+
+		if ( ! forceUpdate ) {
+			if ( areMapsEqual( this.previousSnapshot, updatedStates, this.isStateEqual.bind( this ) ) ) {
+				this.logger.debug( 'Awareness state unchanged, not updating subscribers.' );
```

Comment:
Noting that this currently is filling up the console logs.

Is this needed? Feel like no given how much is in the logs.